src/compiler/checker.ts

```diff
@@ -3643,9 +3630,17 @@ module ts {
         }
 
         function getReducedTypeOfUnionType(type: UnionType): Type {
-            // If union type was created without subtype reduction, perform the deferred reduction now
+            // If union type was created without subtype reduction, perform the deferred reduction now. If a circularity
+            // is detected, simply use the type itself.
             if (!type.reducedType) {
-                type.reducedType = getUnionType(type.types, /*noSubtypeReduction*/ false);
+                type.reducedType = resolvingType;
+                let reducedType = getUnionType(type.types, /*noSubtypeReduction*/ false);
+                if (type.reducedType === resolvingType) {
+                    type.reducedType = reducedType;
+                }
+            }
+            else if (type.reducedType === resolvingType) {
+                type.reducedType = type;

```

Comment:
1. No error when we hit the circularity?  Or is that handled elsewhere?  If handled elsewhere, can you comment in the code where.
2.  It seems like you're setting the reduced type to the type itself.  That seems... odd.  What's the intuition for why that's the appropriate type.